File: session_finder/parsers.py
"""Streaming parsers for the local AI session data sources."""

from __future__ import annotations

import json
import re
import sqlite3
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Iterator
from urllib.parse import quote, unquote


TOOLS = ("opencode", "grok", "codex", "kimi-code", "claude")
NOISE_PREFIXES = (
    "<supermemory",
    "[SUPERMEMORY",
    "<system-reminder>",
    "<user_info>",
)
UUID_RE = re.compile(
    r"[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}"
)
WORKSPACE_PATH_RE = re.compile(r"(?:Workspace Path|工作区路径):\s*(.+)")


@dataclass(frozen=True)
class MessageRecord:
    """One normalized message emitted by a source parser."""

    tool: str
    session_id: str
    cwd: str
    title: str
    timestamp: Any
    role: str
    text: str
    source_path: str


@dataclass(frozen=True)
class SourceSpec:
    """A source file and optional metadata files used to parse it."""

    tool: str
    path: Path
    auxiliary_paths: tuple[Path, ...] = ()


def extract_text(value: Any) -> str:
    """Extract textual content from strings and common message block lists."""

    if isinstance(value, str):
        return value
    if isinstance(value, list):
        parts: list[str] = []
        for item in value:
            if isinstance(item, str):
                parts.append(item)
            elif isinstance(item, dict):
                text = item.get("text")
                if isinstance(text, str):
                    parts.append(text)
                elif "content" in item:
                    nested = extract_text(item.get("content"))
                    if nested:
                        parts.append(nested)
        return "\n".join(part for part in parts if part)
    if isinstance(value, dict):
        text = value.get("text")
        if isinstance(text, str):
            return text
        content = value.get("content")
        if content is not None:
            return extract_text(content)
    return ""


def is_noise(text: str) -> bool:
    """Return whether text is a known injected/system noise record."""

    return text.lstrip().startswith(NOISE_PREFIXES)


def normalize_role(role: Any, text: str) -> str:
    """Normalize source roles and classify injected records as system."""

    if is_noise(text):
        return "system"
    normalized = str(role or "").lower()
    if normalized in {"user", "assistant"}:
        return normalized
    return "system"


def title_from_text(text: str, limit: int = 120) -> str:
    """Create a compact fallback title from a user message."""

    first_line = text.strip().splitlines()[0] if text.strip() else ""
    first_line = re.sub(r"\s+", " ", first_line).strip()
    return first_line[:limit]


def iter_source_specs() -> Iterator[SourceSpec]:
    """Yield existing source files in deterministic tool order."""

    home = Path.home()

    opencode_db = home / ".local" / "share" / "opencode" / "opencode.db"
    if opencode_db.is_file():
        yield SourceSpec("opencode", opencode_db)

    grok_root = home / ".grok" / "sessions"
    if grok_root.is_dir():
        for chat_path in sorted(grok_root.rglob("chat_history.jsonl")):
            if chat_path.is_file():
                summary_path = chat_path.parent / "summary.json"
                auxiliary = (summary_path,) if summary_path.is_file() else ()
                yield SourceSpec("grok", chat_path, auxiliary)

    codex_root = home / ".codex" / "sessions"
    if codex_root.is_dir():
        for rollout_path in sorted(codex_root.rglob("*.jsonl")):
            if rollout_path.is_file():
                yield SourceSpec("codex", rollout_path)

    kimi_root = home / ".kimi-code" / "sessions"
    if kimi_root.is_dir():
        for wire_path in sorted(kimi_root.rglob("wire.jsonl")):
            if not wire_path.is_file():
                continue
            if wire_path.parent.parent.name != "agents":
                continue
            if not wire_path.parent.parent.parent.name.startswith("session_"):
                continue
            yield SourceSpec("kimi-code", wire_path)

    claude_root = home / ".claude" / "projects"
    if claude_root.is_dir():
        for transcript_path in sorted(claude_root.glob("*/*.jsonl")):
            if transcript_path.is_file():
                yield SourceSpec("claude", transcript_path)


def iter_records(spec: SourceSpec) -> Iterator[MessageRecord]:
    """Dispatch a source specification to its streaming parser."""

    if spec.tool == "opencode":
        yield from iter_opencode_records(spec.path)
    elif spec.tool == "grok":
        summary_path = spec.auxiliary_paths[0] if spec.auxiliary_paths else None
        yield from iter_grok_records(spec.path, summary_path)
    elif spec.tool == "codex":
        yield from iter_codex_records(spec.path)
    elif spec.tool == "kimi-code":
        yield from iter_kimi_records(spec.path)
    elif spec.tool == "claude":
        yield from iter_claude_records(spec.path)


def _record(
    tool: str,
    session_id: Any,
    cwd: Any,
    title: Any,
    timestamp: Any,
    role: Any,
    text: str,
    source_path: Path,
) -> MessageRecord | None:
    """Build a normalized record, dropping records without usable text."""

    text = text if isinstance(text, str) else str(text or "")
    if not text:
        return None
    session = str(session_id or "").strip()
    if not session:
        return None
    return MessageRecord(
        tool=tool,
        session_id=session,
        cwd=str(cwd or ""),
        title=str(title or ""),
        timestamp=timestamp,
        role=normalize_role(role, text),
        text=text,
        source_path=str(source_path),
    )


def iter_opencode_records(db_path: Path, session_id: str | None = None) -> Iterator[MessageRecord]:
    """Stream text parts from opencode using a read-only SQLite URI.

    When session_id is given, only that session's messages are streamed —
    this powers session-level incremental indexing.
    """

    uri = "file:" + quote(str(db_path), safe="/") + "?mode=ro"
    connection = sqlite3.connect(uri, uri=True)
    try:
        session_filter = "AND s.id = ?" if session_id else ""
        params = (session_id,) if session_id else ()
        query = f"""
            SELECT
                s.id AS session_id,
                s.directory AS cwd,
                s.title AS title,
                m.id AS message_id,
                m.time_created AS message_created,
                json_extract(m.data, '$.role') AS role,
                p.time_created AS part_created,
                json_extract(p.data, '$.text') AS text
            FROM session AS s
            JOIN message AS m ON m.session_id = s.id
            JOIN part AS p ON p.message_id = m.id
            WHERE json_extract(p.data, '$.text') IS NOT NULL
            {session_filter}
            ORDER BY s.id, m.time_created, m.id, p.time_created, p.id
        """
        current_key: tuple[str, str] | None = None
        current: dict[str, Any] | None = None
        text_parts: list[str] = []

        def flush() -> MessageRecord | None:
            if current is None:
                return None
            return _record(
                "opencode",
                current["session_id"],
                current["cwd"],
                current["title"],
                current["message_created"] or current["part_created"],
                current["role"],
                "\n".join(text_parts),
                db_path,
            )

        for row in connection.execute(query, params):
            key = (str(row[0]), str(row[3]))
            if current_key is not None and key != current_key:
                record = flush()
                if record is not None:
                    yield record
                text_parts = []
            if key != current_key:
                current = {
                    "session_id": row[0],
                    "cwd": row[1],
                    "title": row[2],
                    "message_created": row[4],
                    "role": row[5],
                    "part_created": row[6],
                }
                current_key = key
            text = row[7]
            if text is not None:
                text_parts.append(str(text))

        record = flush()
        if record is not None:
            yield record
    finally:
        connection.close()


def _load_summary(summary_path: Path | None) -> dict[str, Any]:
    """Load an optional Grok summary without affecting message streaming."""

    if summary_path is None:
        return {}
    try:
        with summary_path.open("r", encoding="utf-8", errors="replace") as handle:
            value = json.load(handle)
    except (OSError, json.JSONDecodeError):
        return {}
    return value if isinstance(value, dict) else {}


def iter_grok_records(
    chat_path: Path, summary_path: Path | None = None
) -> Iterator[MessageRecord]:
    """Stream user and assistant records from a Grok chat history."""

    summary = _load_summary(summary_path)
    info = summary.get("info") if isinstance(summary.get("info"), dict) else {}
    encoded_cwd = chat_path.parent.parent.name
    cwd = info.get("cwd") or summary.get("cwd") or unquote(encoded_cwd)
    title = summary.get("generated_title") or info.get("generated_title") or ""
    timestamp = summary.get("created_at") or info.get("created_at")
    session_id = chat_path.parent.name

    try:
        handle = chat_path.open("r", encoding="utf-8", errors="replace")
    except OSError:
        return
    with handle:
        for line in handle:
            try:
                value = json.loads(line)
            except json.JSONDecodeError:
                continue
            if not isinstance(value, dict) or value.get("type") not in {
                "user",
                "assistant",
            }:
                continue
            text = extract_text(value.get("content"))
            record = _record(
                "grok",
                session_id,
                cwd,
                title,
                timestamp,
                value.get("type"),
                text,
                chat_path,
            )
            if record is not None:
                yield record


def _filename_session_id(path: Path) -> str:
    """Extract a UUID from a rollout filename when metadata is absent."""

    match = UUID_RE.search(path.name)
    return match.group(0) if match else path.stem


def iter_codex_records(rollout_path: Path) -> Iterator[MessageRecord]:
    """Stream message payloads from a Codex rollout JSONL file."""

    session_id = _filename_session_id(rollout_path)
    cwd = ""
    session_timestamp: Any = None
    title = ""
    try:
        handle = rollout_path.open("r", encoding="utf-8", errors="replace")
    except OSError:
        return
    with handle:
        for line in handle:
            # Cheap pre-filter: most lines are tool calls we never use.
            if '"message"' not in line and "session_meta" not in line:
                continue
            try:
                value = json.loads(line)
            except json.JSONDecodeError:
                continue
            if not isinstance(value, dict):
                continue
            payload = value.get("payload")
            if not isinstance(payload, dict):
                continue
            line_type = value.get("type")
            if line_type == "session_meta":
                session_id = str(payload.get("id") or session_id)
                cwd = str(payload.get("cwd") or cwd)
                session_timestamp = payload.get("timestamp") or value.get("timestamp")
                continue
            if payload.get("type") != "message":
                continue
            text = extract_text(payload.get("content"))
            record = _record(
                "codex",
                session_id,
                cwd,
                title,
                value.get("timestamp") or payload.get("timestamp") or session_timestamp,
                payload.get("role"),
                text,
                rollout_path,
            )
            if record is not None:
                yield record


def _kimi_workspace_path(value: Any, text: str) -> str:
    """Find a Kimi workspace path from event metadata or injected user info."""

    if isinstance(value, dict):
        for key in ("cwd", "directory", "workspace_path"):
            candidate = value.get(key)
            if isinstance(candidate, str) and candidate:
                return candidate
    match = WORKSPACE_PATH_RE.search(text)
    return match.group(1).strip() if match else ""


def iter_kimi_records(wire_path: Path) -> Iterator[MessageRecord]:
    """Stream context.append_message records from a Kimi wire file."""

    session_dir = wire_path.parent.parent.parent
    session_id = session_dir.name
    cwd = ""
    title = ""
    try:
        handle = wire_path.open("r", encoding="utf-8", errors="replace")
    except OSError:
        return
    with handle:
        for line in handle:
            if "context.append_message" not in line:
                continue
            try:
                value = json.loads(line)
            except json.JSONDecodeError:
                continue
            if not isinstance(value, dict) or value.get("type") != "context.append_message":
                continue
            message = value.get("message")
            if not isinstance(message, dict):
                continue
            text = extract_text(message.get("content"))
            if not text:
                continue
            cwd = _kimi_workspace_path(message, text) or cwd
            origin = message.get("origin")
            if not isinstance(origin, dict):
                origin = value.get("origin") if isinstance(value.get("origin"), dict) else {}
            role = message.get("role")
            if origin.get("kind") == "user":
                role = "user"
            record = _record(
                "kimi-code",
                session_id,
                cwd,
                title,
                value.get("time") or message.get("time"),
                role,
                text,
                wire_path,
            )
            if record is not None:
                yield record


def iter_claude_records(transcript_path: Path) -> Iterator[MessageRecord]:
    """Stream user and assistant messages from a Claude transcript."""

    fallback_session_id = transcript_path.stem
    try:
        handle = transcript_path.open("r", encoding="utf-8", errors="replace")
    except OSError:
        return
    with handle:
        for line in handle:
            if '"message"' not in line:
                continue
            try:
                value = json.loads(line)
            except json.JSONDecodeError:
                continue
            if not isinstance(value, dict):
                continue
            message = value.get("message")
            if not isinstance(message, dict):
                continue
            outer_type = value.get("type")
            if outer_type not in {"user", "assistant"}:
                continue
            text = extract_text(message.get("content"))
            session_id = value.get("sessionId") or value.get("session_id") or fallback_session_id
            cwd = value.get("cwd") or value.get("directory") or ""
            timestamp = value.get("timestamp") or message.get("timestamp")
            record = _record(
                "claude",
                session_id,
                cwd,
                value.get("title") or "",
                timestamp,
                message.get("role") or outer_type,
                text,
                transcript_path,
            )
            if record is not None:
                yield record

File: internal/parsers/parsers.go
// Package parsers discovers local AI session stores and emits normalized records.
package parsers

import (
	"bufio"
	"bytes"
	"database/sql"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/url"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"unicode"

	_ "modernc.org/sqlite"

	"github.com/BayInl/session-finder/internal/record"
)

var (
	uuidRE       = regexp.MustCompile(`[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}`)
	workspaceRE  = regexp.MustCompile(`(?:Workspace Path|工作区路径):\s*(.+)`)
	codexJSONLRE = regexp.MustCompile(`"type"\s*:\s*"(?:message|session_meta)"`)
	kimiJSONLRE  = regexp.MustCompile(`"type"\s*:\s*"context\.append_message"`)
)

// Emit receives one normalized message. Returning an error stops parsing.
type Emit func(record.MessageRecord) error

// Discover returns existing sources in deterministic tool order.
func Discover() ([]record.SourceSpec, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return nil, err
	}
	var specs []record.SourceSpec

	opencode := filepath.Join(home, ".local", "share", "opencode", "opencode.db")
	if isFile(opencode) {
		specs = append(specs, record.SourceSpec{Tool: "opencode", Path: opencode})
	}

	grokRoot := filepath.Join(home, ".grok", "sessions")
	for _, path := range matchingFiles(grokRoot, func(path string, d os.DirEntry) bool {
		return d.Name() == "chat_history.jsonl"
	}) {
		summary := filepath.Join(filepath.Dir(path), "summary.json")
		aux := []string{}
		if isFile(summary) {
			aux = append(aux, summary)
		}
		specs = append(specs, record.SourceSpec{Tool: "grok", Path: path, AuxiliaryPath: aux})
	}

	codexRoot := filepath.Join(home, ".codex", "sessions")
	for _, path := range matchingFiles(codexRoot, func(path string, d os.DirEntry) bool {
		return strings.HasSuffix(d.Name(), ".jsonl")
	}) {
		specs = append(specs, record.SourceSpec{Tool: "codex", Path: path})
	}

	kimiRoot := filepath.Join(home, ".kimi-code", "sessions")
	for _, path := range matchingFiles(kimiRoot, func(path string, d os.DirEntry) bool {
		if d.Name() != "wire.jsonl" {
			return false
		}
		agentsDir := filepath.Dir(filepath.Dir(path))
		sessionDir := filepath.Dir(agentsDir)
		return filepath.Base(agentsDir) == "agents" && strings.HasPrefix(filepath.Base(sessionDir), "session_")
	}) {
		specs = append(specs, record.SourceSpec{Tool: "kimi-code", Path: path})
	}

	claudeRoot := filepath.Join(home, ".claude", "projects")
	var claude []string
	if entries, readErr := os.ReadDir(claudeRoot); readErr == nil {
		for _, project := range entries {
			if !project.IsDir() {
				continue
			}
			files, _ := filepath.Glob(filepath.Join(claudeRoot, project.Name(), "*.jsonl"))
			claude = append(claude, files...)
		}
	}
	sort.Strings(claude)
	for _, path := range claude {
		if isFile(path) {
			specs = append(specs, record.SourceSpec{Tool: "claude", Path: path})
		}
	}
	return specs, nil
}

func isFile(path string) bool {
	info, err := os.Stat(path)
	return err == nil && info.Mode().IsRegular()
}

func matchingFiles(root string, predicate func(string, os.DirEntry) bool) []string {
	var paths []string
	if _, err := os.Stat(root); err != nil {
		return paths
	}
	_ = filepath.WalkDir(root, func(path string, d os.DirEntry, err error) error {
		if err != nil {
			return nil
		}
		if d.IsDir() {
			return nil
		}
		if predicate(path, d) && isFile(path) {
			paths = append(paths, path)
		}
		return nil
	})
	sort.Strings(paths)
	return paths
}

// Parse dispatches a source specification to the corresponding parser.
func Parse(spec record.SourceSpec, emit Emit) error {
	switch spec.Tool {
	case "opencode":
		return Opencode(spec.Path, "", emit)
	case "grok":
		var summary string
		if len(spec.AuxiliaryPath) > 0 {
			summary = spec.AuxiliaryPath[0]
		}
		return Grok(spec.Path, summary, emit)
	case "codex":
		return Codex(spec.Path, emit)
	case "kimi-code":
		return Kimi(spec.Path, emit)
	case "claude":
		return Claude(spec.Path, emit)
	default:
		return fmt.Errorf("unknown tool %q", spec.Tool)
	}
}

// ExtractText extracts textual content from strings and common message blocks.
func ExtractText(value any) string {
	switch value := value.(type) {
	case string:
		return value
	case []any:
		var parts []string
		for _, item := range value {
			switch item := item.(type) {
			case string:
				if item != "" {
					parts = append(parts, item)
				}
			case map[string]any:
				if text, ok := item["text"].(string); ok {
					if text != "" {
						parts = append(parts, text)
					}
				} else if nested, ok := item["content"]; ok {
					if text := ExtractText(nested); text != "" {
						parts = append(parts, text)
					}
				}
			}
		}
		return strings.Join(parts, "\n")
	case map[string]any:
		if text, ok := value["text"].(string); ok {
			return text
		}
		if content, ok := value["content"]; ok {
			return ExtractText(content)
		}
	}
	return ""
}

// IsNoise identifies known injected/system noise records.
func IsNoise(text string) bool {
	trimmed := strings.TrimLeftFunc(text, unicode.IsSpace)
	for _, prefix := range record.NoisePrefixes {
		if strings.HasPrefix(trimmed, prefix) {
			return true
		}
	}
	return false
}

// NormalizeRole maps source roles to user/assistant/system.
func NormalizeRole(role any, text string) string {
	if IsNoise(text) {
		return "system"
	}
	normalized := strings.ToLower(anyString(role))
	if normalized == "user" || normalized == "assistant" {
		return normalized
	}
	return "system"
}

// TitleFromText creates a compact fallback title from a user message.
func TitleFromText(text string, limit int) string {
	if limit <= 0 {
		return ""
	}
	trimmed := strings.TrimSpace(text)
	if trimmed == "" {
		return ""
	}
	first := strings.Split(trimmed, "\n")[0]
	first = strings.Join(strings.Fields(first), " ")
	runes := []rune(first)
	if len(runes) > limit {
		runes = runes[:limit]
	}
	return string(runes)
}

func anyString(value any) string {
	if value == nil {
		return ""
	}
	if s, ok := value.(string); ok {
		return s
	}
	return fmt.Sprint(value)
}

func makeRecord(tool string, sessionID, cwd, title, timestamp, role any, text, sourcePath string) (record.MessageRecord, bool) {
	if text == "" {
		text = anyString(text)
	}
	if text == "" {
		return record.MessageRecord{}, false
	}
	session := strings.TrimSpace(anyString(sessionID))
	if session == "" {
		return record.MessageRecord{}, false
	}
	return record.MessageRecord{
		Tool:       tool,
		SessionID:  session,
		CWD:        anyString(cwd),
		Title:      anyString(title),
		Timestamp:  timestamp,
		Role:       NormalizeRole(role, text),
		Text:       text,
		SourcePath: sourcePath,
	}, true
}

func fileURI(path string, readOnly bool) string {
	// SQLite file URIs use a slash-preserving escaped absolute path.
	escaped := url.PathEscape(path)
	escaped = strings.ReplaceAll(escaped, "%2F", "/")
	escaped = strings.ReplaceAll(escaped, "%2f", "/")
	uri := "file:" + escaped
	if readOnly {
		uri += "?mode=ro"
	}
	return uri
}

// Opencode streams text parts from an opencode SQLite store. If sessionID is
// non-empty, only that session's messages are emitted.
func Opencode(dbPath, sessionID string, emit Emit) error {
	db, err := sql.Open("sqlite", fileURI(dbPath, true))
	if err != nil {
		return err
	}
	defer db.Close()
	filter := ""
	args := []any{}
	if sessionID != "" {
		filter = " AND s.id = ?"
		args = append(args, sessionID)
	}
	query := `
		SELECT
			s.id AS session_id,
			s.directory AS cwd,
			s.title AS title,
			m.id AS message_id,
			m.time_created AS message_created,
			json_extract(m.data, '$.role') AS role,
			p.time_created AS part_created,
			json_extract(p.data, '$.text') AS text
		FROM session AS s
		JOIN message AS m ON m.session_id = s.id
		JOIN part AS p ON p.message_id = m.id
		WHERE json_extract(p.data, '$.text') IS NOT NULL` + filter + `
		ORDER BY s.id, m.time_created, m.id, p.time_created, p.id`
	rows, err := db.Query(query, args...)
	if err != nil {
		return err
	}
	defer rows.Close()

	var currentKey string
	var currentSession, currentCWD, currentTitle, currentRole string
	var messageCreated, partCreated any
	var textParts []string
	flush := func() error {
		if currentKey == "" {
			return nil
		}
		timestamp := messageCreated
		if isEmptyValue(timestamp) {
			timestamp = partCreated
		}
		value, ok := makeRecord("opencode", currentSession, currentCWD, currentTitle, timestamp, currentRole, strings.Join(textParts, "\n"), dbPath)
		if !ok {
			return nil
		}
		return emit(value)
	}
	for rows.Next() {
		var sid, cwd, title, messageID, role, text sql.NullString
		var msgCreated, pCreated any
		if err := rows.Scan(&sid, &cwd, &title, &messageID, &msgCreated, &role, &pCreated, &text); err != nil {
			return err
		}
		key := sid.String + "\x00" + messageID.String
		if currentKey != "" && key != currentKey {
			if err := flush(); err != nil {
				return err
			}
			textParts = nil
		}
		if key != currentKey {
			currentKey = key
			currentSession = sid.String
			currentCWD = cwd.String
			currentTitle = title.String
			currentRole = role.String
			messageCreated = msgCreated
			partCreated = pCreated
		}
		if text.Valid {
			textParts = append(textParts, text.String)
		}
	}
	if err := rows.Err(); err != nil {
		return err
	}
	return flush()
}

func isEmptyValue(value any) bool {
	if value == nil {
		return true
	}
	switch value := value.(type) {
	case string:
		return value == ""
	case []byte:
		return len(value) == 0
	case bool:
		return !value
	case int:
		return value == 0
	case int32:
		return value == 0
	case int64:
		return value == 0
	case float32:
		return value == 0
	case float64:
		return value == 0
	case sql.NullString:
		return !value.Valid || value.String == ""
	case sql.NullInt64:
		return !value.Valid || value.Int64 == 0
	}
	return false
}

func loadJSON(path string) (map[string]any, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	var value map[string]any
	if err := json.Unmarshal(data, &value); err != nil {
		return nil, err
	}
	return value, nil
}

func loadSummary(path string) map[string]any {
	if path == "" {
		return map[string]any{}
	}
	value, err := loadJSON(path)
	if err != nil {
		return map[string]any{}
	}
	return value
}

func nestedMap(value any) map[string]any {
	if result, ok := value.(map[string]any); ok {
		return result
	}
	return map[string]any{}
}

func firstNonEmpty(values ...any) any {
	for _, value := range values {
		if !isEmptyValue(value) {
			return value
		}
	}
	return nil
}

// Grok streams user and assistant records from a Grok chat history.
func Grok(chatPath, summaryPath string, emit Emit) error {
	summary := loadSummary(summaryPath)
	info := nestedMap(summary["info"])
	encodedCWD := filepath.Base(filepath.Dir(filepath.Dir(chatPath)))
	cwd := firstNonEmpty(info["cwd"], summary["cwd"])
	if isEmptyValue(cwd) {
		cwd, _ = url.PathUnescape(encodedCWD)
	}
	title := firstNonEmpty(summary["generated_title"], info["generated_title"])
	timestamp := firstNonEmpty(summary["created_at"], info["created_at"])
	sessionID := filepath.Base(filepath.Dir(chatPath))
	return eachJSONL(chatPath, nil, nil, func(value map[string]any) error {
		lineType, _ := value["type"].(string)
		if lineType != "user" && lineType != "assistant" {
			return nil
		}
		text := ExtractText(value["content"])
		if result, ok := makeRecord("grok", sessionID, cwd, title, timestamp, lineType, text, chatPath); ok {
			return emit(result)
		}
		return nil
	})
}

func filenameSessionID(path string) string {
	name := filepath.Base(path)
	if match := uuidRE.FindString(name); match != "" {
		return match
	}
	return strings.TrimSuffix(name, filepath.Ext(name))
}

// Codex streams message payloads from a Codex rollout JSONL file.
func Codex(rolloutPath string, emit Emit) error {
	sessionID := filenameSessionID(rolloutPath)
	cwd := ""
	var sessionTimestamp any
	return eachJSONL(rolloutPath, codexJSONLRE, [][]byte{[]byte(`"message"`), []byte(`session_meta`)}, func(value map[string]any) error {
		payload := nestedMap(value["payload"])
		if len(payload) == 0 {
			return nil
		}
		lineType, _ := value["type"].(string)
		if lineType == "session_meta" {
			if id := anyString(payload["id"]); id != "" {
				sessionID = id
			}
			if value := anyString(payload["cwd"]); value != "" {
				cwd = value
			}
			sessionTimestamp = firstNonEmpty(payload["timestamp"], value["timestamp"])
			return nil
		}
		if anyString(payload["type"]) != "message" {
			return nil
		}
		text := ExtractText(payload["content"])
		timestamp := firstNonEmpty(value["timestamp"], payload["timestamp"], sessionTimestamp)
		if result, ok := makeRecord("codex", sessionID, cwd, "", timestamp, payload["role"], text, rolloutPath); ok {
			return emit(result)
		}
		return nil
	})
}

func kimiWorkspacePath(value map[string]any, text string) string {
	for _, key := range []string{"cwd", "directory", "workspace_path"} {
		if candidate, ok := value[key].(string); ok && candidate != "" {
			return candidate
		}
	}
	if !strings.Contains(text, "Workspace Path") && !strings.Contains(text, "工作区路径") {
		return ""
	}
	match := workspaceRE.FindStringSubmatch(text)
	if len(match) > 1 {
		return strings.TrimSpace(match[1])
	}
	return ""
}

// Kimi streams context.append_message records from a Kimi wire file.
func Kimi(wirePath string, emit Emit) error {
	sessionDir := filepath.Dir(filepath.Dir(filepath.Dir(wirePath)))
	sessionID := filepath.Base(sessionDir)
	cwd := ""
	return eachJSONL(wirePath, kimiJSONLRE, [][]byte{[]byte(`context.append_message`)}, func(value map[string]any) error {
		if anyString(value["type"]) != "context.append_message" {
			return nil
		}
		message := nestedMap(value["message"])
		if len(message) == 0 {
			return nil
		}
		text := ExtractText(message["content"])
		if text == "" {
			return nil
		}
		if candidate := kimiWorkspacePath(message, text); candidate != "" {
			cwd = candidate
		}
		origin := nestedMap(message["origin"])
		if len(origin) == 0 {
			origin = nestedMap(value["origin"])
		}
		role := message["role"]
		if anyString(origin["kind"]) == "user" {
			role = "user"
		}
		timestamp := firstNonEmpty(value["time"], message["time"])
		if result, ok := makeRecord("kimi-code", sessionID, cwd, "", timestamp, role, text, wirePath); ok {
			return emit(result)
		}
		return nil
	})
}

// Claude streams user and assistant messages from a Claude transcript.
func Claude(transcriptPath string, emit Emit) error {
	fallback := strings.TrimSuffix(filepath.Base(transcriptPath), filepath.Ext(transcriptPath))
	return eachJSONL(transcriptPath, nil, [][]byte{[]byte(`"message"`)}, func(value map[string]any) error {
		message := nestedMap(value["message"])
		if len(message) == 0 {
			return nil
		}
		outerType := anyString(value["type"])
		if outerType != "user" && outerType != "assistant" {
			return nil
		}
		text := ExtractText(message["content"])
		sessionID := anyString(firstNonEmpty(value["sessionId"], value["session_id"]))
		if sessionID == "" {
			sessionID = fallback
		}
		cwd := firstNonEmpty(value["cwd"], value["directory"])
		timestamp := firstNonEmpty(value["timestamp"], message["timestamp"])
		title := value["title"]
		role := firstNonEmpty(message["role"], outerType)
		if result, ok := makeRecord("claude", sessionID, cwd, title, timestamp, role, text, transcriptPath); ok {
			return emit(result)
		}
		return nil
	})
}

func eachJSONL(path string, candidate *regexp.Regexp, hints [][]byte, emit func(map[string]any) error) error {
	handle, err := os.Open(path)
	if err != nil {
		return err
	}
	defer handle.Close()
	reader := bufio.NewReaderSize(handle, 256*1024)
	for {
		raw, readErr := reader.ReadBytes('\n')
		if len(raw) > 0 {
			if len(hints) > 0 {
				matched := false
				for _, hint := range hints {
					if bytes.Contains(raw, hint) {
						matched = true
						break
					}
				}
				if !matched {
					if readErr != nil && !errors.Is(readErr, io.EOF) {
						return readErr
					}
					if readErr != nil {
						break
					}
					continue
				}
			}
			if candidate != nil && !candidate.Match(raw) {
				if readErr != nil && !errors.Is(readErr, io.EOF) {
					return readErr
				}
				if readErr != nil {
					break
				}
				continue
			}
			var value map[string]any
			if json.Unmarshal(bytes.TrimSpace(raw), &value) == nil && value != nil {
				if err := emit(value); err != nil {
					return err
				}
			}
		}
		if readErr != nil {
			if !errors.Is(readErr, io.EOF) {
				return readErr
			}
			break
		}
	}
	return nil
}
